File: extension/test/historyMerge.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { mergeMillisByDate, mergeHistories } from '../src/lib/historyMerge.js';
import { isPerfectDay } from '../src/lib/gamification.js';

// The dashboard used to draw only this device's chrome.storage history, so a reinstall or a
// second device showed an empty heatmap. These cover the merge rule that fills that hole:
// per-date Math.max(local, server), because either side can be the one that is behind.

const row = (date, usage_ms, shorts_ms = 0, emergency_ms = 0, emergency_uses = 0) => ({
  date,
  usage_ms,
  shorts_ms,
  emergency_ms,
  emergency_uses
});

/** emergency_uses 컬럼이 생기기 전에 쓰인 행 — 그 칸이 통째로 비어 있다. */
const legacyRow = (date, usage_ms, emergency_ms = 0) => ({
  date,
  usage_ms,
  shorts_ms: 0,
  emergency_ms
});

test('a day only the server knows about shows up in the merged history', () => {
  const merged = mergeMillisByDate({}, [row('2026-09-01', 1_800_000)], 'usage_ms');
  assert.deepEqual(merged, { '2026-09-01': 1_800_000 });
});

test('a day only the local device knows about survives the merge', () => {
  const merged = mergeMillisByDate({ '2026-09-01': 600_000 }, [], 'usage_ms');
  assert.deepEqual(merged, { '2026-09-01': 600_000 });
});

test('the bigger of the two wins when the local figure is ahead', () => {
  // Not yet synced: the service worker throttles daily_usage pushes to 30s.
  const merged = mergeMillisByDate(
    { '2026-09-01': 900_000 },
    [row('2026-09-01', 600_000)],
    'usage_ms'
  );
  assert.deepEqual(merged, { '2026-09-01': 900_000 });
});

test('the bigger of the two wins when the server figure is ahead', () => {
  // The usual case: the server row also carries the other devices' share.
  const merged = mergeMillisByDate(
    { '2026-09-01': 600_000 },
    [row('2026-09-01', 900_000)],
    'usage_ms'
  );
  assert.deepEqual(merged, { '2026-09-01': 900_000 });
});

test('the two sides are never summed, which would double count this device', () => {
  const merged = mergeMillisByDate(
    { '2026-09-01': 600_000 },
    [row('2026-09-01', 600_000)],
    'usage_ms'
  );
  assert.deepEqual(merged, { '2026-09-01': 600_000 });
});

test('empty input on both sides merges to an empty map', () => {
  assert.deepEqual(mergeMillisByDate({}, [], 'usage_ms'), {});
  assert.deepEqual(mergeMillisByDate(undefined, undefined, 'usage_ms'), {});
  assert.deepEqual(mergeMillisByDate(null, null, 'usage_ms'), {});
});

test('shorts use the same rule against their own column', () => {
  const rows = [row('2026-09-01', 900_000, 300_000)];
  assert.deepEqual(mergeMillisByDate({ '2026-09-01': 120_000 }, rows, 'shorts_ms'), {
    '2026-09-01': 300_000
  });
});

test('bigint columns arriving as strings are still compared as numbers', () => {
  // PostgREST can hand back bigint as a string; '900000' < '600000' lexicographically.
  const merged = mergeMillisByDate(
    { '2026-09-01': 600_000 },
    [row('2026-09-01', '900000')],
    'usage_ms'
  );
  assert.deepEqual(merged, { '2026-09-01': 900_000 });
});

test('missing or junk values are normalised to zero instead of poisoning the max', () => {
  const rows = [row('2026-09-01', null), { date: '2026-09-02' }, null, { usage_ms: 5 }];
  const merged = mergeMillisByDate({ '2026-09-01': -5, '2026-09-03': 'x' }, rows, 'usage_ms');
  assert.deepEqual(merged, { '2026-09-01': 0, '2026-09-02': 0, '2026-09-03': 0 });
});

test('all three histories merge together in one pass', () => {
  const local = {
    usage: { '2026-09-01': 900_000 },
    shorts: { '2026-09-01': 120_000 },
    emergency: { '2026-09-01': { uses: 2, ms: 300_000 } }
  };
  const rows = [
    row('2026-09-01', 600_000, 400_000, 200_000),
    row('2026-09-02', 1_200_000, 0, 60_000)
  ];
  const merged = mergeHistories(local, rows);

  assert.deepEqual(merged.usage, { '2026-09-01': 900_000, '2026-09-02': 1_200_000 });
  assert.deepEqual(merged.shorts, { '2026-09-01': 400_000, '2026-09-02': 0 });
  assert.equal(merged.emergency['2026-09-01'].ms, 300_000);
  assert.equal(merged.emergency['2026-09-02'].ms, 60_000);
});

test('emergency uses take the bigger of the local record and the server column', () => {
  const local = {
    usage: { '2026-09-01': 900_000 },
    emergency: { '2026-09-01': { uses: 3, ms: 200_000 } }
  };
  // Not yet pushed: the local count is ahead of the server's.
  const merged = mergeHistories(local, [row('2026-09-01', 900_000, 0, 500_000, 1)]);
  assert.deepEqual(merged.emergency['2026-09-01'], { ms: 500_000, uses: 3 });
});

test('a day only the server knows about keeps the count the server reports', () => {
  const merged = mergeHistories({ usage: {}, emergency: {} }, [
    row('2026-09-05', 900_000, 0, 120_000, 2)
  ]);
  assert.deepEqual(merged.emergency['2026-09-05'], { ms: 120_000, uses: 2 });
});

test('an emergency the phone was granted but never watched is not a perfect day', () => {
  // The regression this column exists for: emergency_ms is 0 because the grant went unused, so
  // merging time alone left uses at null -> isPerfectDay(..., uses ?? 0) painted the day perfect.
  // Android's sync/HistoryMerge.kt already merges this column; the two must not diverge.
  const merged = mergeHistories({ usage: {}, emergency: {} }, [
    row('2026-09-05', 600_000, 0, 0, 1)
  ]);
  const day = merged.emergency['2026-09-05'];
  assert.deepEqual(day, { ms: 0, uses: 1 });
  assert.equal(isPerfectDay(merged.usage['2026-09-05'], 1_800_000, day.ms, day.uses ?? 0), false);
});

test('a server row from before the column existed still means "unknown", not zero', () => {
  const merged = mergeHistories({ usage: {}, emergency: {} }, [
    legacyRow('2026-09-05', 900_000, 120_000)
  ]);
  assert.deepEqual(merged.emergency['2026-09-05'], { ms: 120_000, uses: null });
  // An explicit null on the row means the same thing.
  const withNull = mergeHistories({ usage: {}, emergency: {} }, [
    { date: '2026-09-06', usage_ms: 1, emergency_uses: null }
  ]);
  assert.equal(withNull.emergency['2026-09-06'].uses, null);
});

test('a locally known day with no server count keeps its local count', () => {
  const local = { usage: { '2026-09-01': 600_000 }, emergency: {} };
  const merged = mergeHistories(local, [legacyRow('2026-09-01', 600_000)]);
  assert.deepEqual(merged.emergency['2026-09-01'], { ms: 0, uses: 0 });
});

test('junk counts from the server never poison the merged count', () => {
  const local = { usage: { '2026-09-01': 600_000 }, emergency: {} };
  const rows = [
    { date: '2026-09-01', usage_ms: 600_000, emergency_uses: 'nonsense' },
    { date: '2026-09-02', usage_ms: 600_000, emergency_uses: -4 },
    // PostgREST can hand back an integer column as a string.
    { date: '2026-09-03', usage_ms: 600_000, emergency_uses: '2' }
  ];
  const merged = mergeHistories(local, rows);
  assert.equal(merged.emergency['2026-09-01'].uses, 0);
  assert.equal(merged.emergency['2026-09-02'].uses, 0);
  assert.equal(merged.emergency['2026-09-03'].uses, 2);
});

test('a locally recorded day with no emergency entry counts as zero uses, not unknown', () => {
  const merged = mergeHistories({ usage: { '2026-09-01': 600_000 }, emergency: {} }, []);
  assert.deepEqual(merged.emergency['2026-09-01'], { ms: 0, uses: 0 });
});

test('every day present in the merged usage also has an emergency entry', () => {
  const local = {
    usage: { '2026-09-01': 600_000 },
    emergency: { '2026-08-30': { uses: 1, ms: 10 } }
  };
  const merged = mergeHistories(local, [row('2026-09-02', 600_000)]);
  Object.keys(merged.usage).forEach((date) => {
    assert.ok(merged.emergency[date], `missing emergency entry for ${date}`);
  });
});

test('merging nothing at all yields three empty maps', () => {
  assert.deepEqual(mergeHistories({}, []), { usage: {}, shorts: {}, emergency: {} });
  assert.deepEqual(mergeHistories(undefined, undefined), { usage: {}, shorts: {}, emergency: {} });
});

File: android/app/src/test/java/com/tubelimiter/app/sync/HistoryMergeTest.kt
package com.tubelimiter.app.sync

import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * 대시보드는 원래 이 기기의 DataStore 기록만 그려서, 앱을 재설치하거나 다른 기기에서만 본 날은
 * 히트맵이 통째로 비었다. 그 구멍을 메우는 병합 규칙 — 날짜별 max(로컬, 서버) — 를 고정한다.
 * 어느 쪽이든 뒤처져 있을 수 있으므로 한쪽을 무조건 채택하거나 둘을 더하면 안 된다.
 */
class HistoryMergeTest {

    private fun row(
        date: String,
        usageMs: Long = 0L,
        emergencyMs: Long = 0L,
        emergencyUses: Int = 0,
    ) = RemoteDailyUsageRow(date, usageMs, emergencyMs, emergencyUses)

    @Test
    fun `서버만 아는 날도 병합 결과에 나온다`() {
        val merged = mergeMillisByDate(emptyMap(), listOf(row("2026-09-01", usageMs = 1_800_000))) { it.usageMs }
        assertEquals(mapOf("2026-09-01" to 1_800_000L), merged)
    }

    @Test
    fun `로컬만 아는 날은 병합 후에도 살아남는다`() {
        val merged = mergeMillisByDate(mapOf("2026-09-01" to 600_000L), emptyList()) { it.usageMs }
        assertEquals(mapOf("2026-09-01" to 600_000L), merged)
    }

    @Test
    fun `로컬이 앞서 있으면 로컬이 이긴다`() {
        // 아직 동기화 전: refreshDailyUsageSync는 30초 스로틀이다.
        val merged = mergeMillisByDate(
            mapOf("2026-09-01" to 900_000L),
            listOf(row("2026-09-01", usageMs = 600_000)),
        ) { it.usageMs }
        assertEquals(mapOf("2026-09-01" to 900_000L), merged)
    }

    @Test
    fun `서버가 앞서 있으면 서버가 이긴다`() {
        // 흔한 경우: 서버 행에는 다른 기기 몫까지 들어 있다.
        val merged = mergeMillisByDate(
            mapOf("2026-09-01" to 600_000L),
            listOf(row("2026-09-01", usageMs = 900_000)),
        ) { it.usageMs }
        assertEquals(mapOf("2026-09-01" to 900_000L), merged)
    }

    @Test
    fun `두 값을 더하지 않는다 - 더하면 이 기기 몫을 두 번 센다`() {
        val merged = mergeMillisByDate(
            mapOf("2026-09-01" to 600_000L),
            listOf(row("2026-09-01", usageMs = 600_000)),
        ) { it.usageMs }
        assertEquals(mapOf("2026-09-01" to 600_000L), merged)
    }

    @Test
    fun `양쪽이 비면 빈 맵이다`() {
        assertEquals(emptyMap<String, Long>(), mergeMillisByDate(emptyMap(), emptyList()) { it.usageMs })
    }

    @Test
    fun `이상값은 max를 오염시키지 않고 0으로 정규화된다`() {
        val merged = mergeMillisByDate(
            mapOf("2026-09-01" to -5L),
            listOf(row("2026-09-01", usageMs = -1), row("", usageMs = 100)),
        ) { it.usageMs }
        assertEquals(mapOf("2026-09-01" to 0L), merged)
    }

    @Test
    fun `긴급 시청 시간도 자기 컬럼으로 같은 규칙을 쓴다`() {
        val merged = mergeMillisByDate(
            mapOf("2026-09-01" to 120_000L),
            listOf(row("2026-09-01", usageMs = 900_000, emergencyMs = 300_000)),
        ) { it.emergencyMs }
        assertEquals(mapOf("2026-09-01" to 300_000L), merged)
    }

    @Test
    fun `서버에만 있는 날은 긴급 횟수도 서버 값을 쓴다`() {
        // 확장과 의도적으로 다른 지점: 확장 historyMerge.js는 이 경우 횟수를 null로 두지만
        // (emergency_uses가 확장 대시보드 경로에 아직 안 붙어 있어서), 안드로이드는 그 컬럼을
        // 같이 읽어 오므로 실제 횟수를 알 수 있다. 그래야 다른 기기에서 긴급 시청을 쓴 날이
        // 완벽한 날에서 제대로 걸러진다.
        val merged = mergeHistories(
            localUsage = emptyMap(),
            localEmergencyMillis = emptyMap(),
            localEmergencyUses = emptyMap(),
            serverRows = listOf(row("2026-09-01", usageMs = 900_000, emergencyMs = 300_000, emergencyUses = 2)),
        )
        assertEquals(900_000L, merged.usage["2026-09-01"])
        assertEquals(MergedEmergencyDay(millis = 300_000L, uses = 2), merged.emergency["2026-09-01"])
    }

    @Test
    fun `긴급 횟수도 날짜별 max다 - 아직 안 올린 로컬 횟수가 지워지지 않는다`() {
        val merged = mergeHistories(
            localUsage = mapOf("2026-09-01" to 900_000L),
            localEmergencyMillis = mapOf("2026-09-01" to 300_000L),
            localEmergencyUses = mapOf("2026-09-01" to 3),
            serverRows = listOf(row("2026-09-01", usageMs = 600_000, emergencyMs = 120_000, emergencyUses = 1)),
        )
        assertEquals(900_000L, merged.usage["2026-09-01"])
        assertEquals(MergedEmergencyDay(millis = 300_000L, uses = 3), merged.emergency["2026-09-01"])
    }

    @Test
    fun `사용량 기록이 있는 날은 긴급 항목이 반드시 만들어진다`() {
        // "항목이 없는 날"과 "긴급 시청이 0회인 날"을 호출부가 헷갈리지 않게.
        val merged = mergeHistories(
            localUsage = mapOf("2026-09-01" to 600_000L),
            localEmergencyMillis = emptyMap(),
            localEmergencyUses = emptyMap(),
            serverRows = emptyList(),
        )
        assertEquals(MergedEmergencyDay(millis = 0L, uses = 0), merged.emergency["2026-09-01"])
    }

    @Test
    fun `서버 조회 실패(빈 목록)면 로컬 기록 그대로 그린다`() {
        val merged = mergeHistories(
            localUsage = mapOf("2026-09-01" to 600_000L, "2026-09-02" to 60_000L),
            localEmergencyMillis = mapOf("2026-09-01" to 30_000L),
            localEmergencyUses = mapOf("2026-09-01" to 1),
            serverRows = emptyList(),
        )
        assertEquals(mapOf("2026-09-01" to 600_000L, "2026-09-02" to 60_000L), merged.usage)
        assertEquals(MergedEmergencyDay(millis = 30_000L, uses = 1), merged.emergency["2026-09-01"])
        assertEquals(MergedEmergencyDay(millis = 0L, uses = 0), merged.emergency["2026-09-02"])
    }

    @Test
    fun `로컬에만 있는 날과 서버에만 있는 날이 한 결과에 모인다`() {
        val merged = mergeHistories(
            localUsage = mapOf("2026-09-01" to 600_000L),
            localEmergencyMillis = emptyMap(),
            localEmergencyUses = emptyMap(),
            serverRows = listOf(row("2026-09-02", usageMs = 900_000)),
        )
        assertEquals(mapOf("2026-09-01" to 600_000L, "2026-09-02" to 900_000L), merged.usage)
    }
}
